File: exercise/kotlin/day20/src/main/kotlin/domain/yahtzee/YahtzeeCalculator.kt
package domain.yahtzee

object YahtzeeCalculator {
    private const val ROLL_LENGTH = 5
    private const val MINIMUM_DIE = 1
    private const val MAXIMUM_DIE = 6

    fun number(dice: IntArray, number: Int): Int = calculate(
        { d -> d.filter { die -> die == number }.sum() }, dice
    )

    fun threeOfAKind(dice: IntArray): Int = calculateNOfAKind(dice, 3)
    fun fourOfAKind(dice: IntArray): Int = calculateNOfAKind(dice, 4)
    fun yahtzee(dice: IntArray): Int = calculate(
        { d -> if (hasNOfAKind(d, 5)) Scores.YAHTZEE_SCORE else 0 }, dice
    )

    private fun calculateNOfAKind(dice: IntArray, n: Int): Int = calculate(
        { d -> if (hasNOfAKind(d, n)) d.sum() else 0 }, dice
    )

    fun fullHouse(dice: IntArray): Int = calculate(
        { d ->
            groupDieByFrequency(d)
                .let { dieFrequency ->
                    if (dieFrequency.containsValue(3) && dieFrequency.containsValue(2)) Scores.HOUSE_SCORE
                    else 0
                }
        }, dice
    )

    fun largeStraight(dice: IntArray): Int = calculate({ d ->
        if (d.sorted()
                .windowed(2)
                .all { pair -> pair[0] + 1 == pair[1] }
        ) Scores.LARGE_STRAIGHT_SCORE else 0
    }, dice)

    fun smallStraight(dice: IntArray): Int = calculate({ d ->
        toSortedString(d)
            .let { diceString -> if (isSmallStraight(diceString)) 30 else 0 }
    }, dice)

    private fun isSmallStraight(diceString: String): Boolean =
        diceString.contains("1234") || diceString.contains("2345") || diceString.contains("3456")

    private fun hasNOfAKind(dice: IntArray, n: Int): Boolean = groupDieByFrequency(dice)
        .values
        .any { count -> count >= n }

    private fun toSortedString(dice: IntArray): String = dice.sorted()
        .distinct()
        .joinToString("")

    fun chance(dice: IntArray): Int = calculate(
        { d -> d.sum() }, dice
    )

    private fun groupDieByFrequency(dice: IntArray): Map<Int, Int> = dice.toList()
        .groupingBy { x -> x }
        .eachCount()

    private fun calculate(compute: (dice: IntArray) -> Int, dice: IntArray): Int {
        dice.validateRoll()
        return compute(dice)
    }

    private fun IntArray.validateRoll() {
        require(!hasInvalidLength()) { "Invalid dice... A roll should contain 6 dice." }
        require(!containsInvalidDie()) { "Invalid die value. Each die must be between 1 and 6." }
    }

    private fun IntArray.hasInvalidLength(): Boolean = size != ROLL_LENGTH
    private fun IntArray.containsInvalidDie(): Boolean = any { die -> die.isValid() }
    private fun Int.isValid(): Boolean = this < MINIMUM_DIE || this > MAXIMUM_DIE

    private object Scores {
        const val YAHTZEE_SCORE = 50
        const val HOUSE_SCORE = 25
        const val LARGE_STRAIGHT_SCORE = 40
    }
}

File: solution/c#/Day20/Day20/Domain/Yahtzee/Hollywood.Principle/YahtzeeCalculator.cs
namespace Day20.Domain.Yahtzee.Hollywood.Principle
{
    public static class YahtzeeCalculator
    {
        private const int RollLength = 5;
        private const int MinimumDie = 1;
        private const int MaximumDie = 6;

        public static void Number(
            int[] dice,
            int number,
            Action<int> onSuccess,
            Action<string> onError)
            => Calculate(d => d.Where(die => die == number).Sum(), dice, onSuccess, onError);

        public static void ThreeOfAKind(
            int[] dice,
            Action<int> onSuccess,
            Action<string> onError) => CalculateNOfAKind(dice, 3, onSuccess, onError);

        public static void FourOfAKind(
            int[] dice,
            Action<int> onSuccess,
            Action<string> onError) => CalculateNOfAKind(dice, 4, onSuccess, onError);

        public static void Yahtzee(
            int[] dice,
            Action<int> onSuccess,
            Action<string> onError)
            => Calculate(d =>
                    HasNOfAKind(d, 5)
                        ? Scores.YahtzeeScore
                        : 0,
                dice, onSuccess, onError);

        private static void CalculateNOfAKind(
            int[] dice,
            int n,
            Action<int> onSuccess,
            Action<string> onError)
            => Calculate(d => HasNOfAKind(d, n) ? d.Sum() : 0, dice, onSuccess, onError);

        public static void FullHouse(
            int[] dice,
            Action<int> onSuccess,
            Action<string> onError)
            => Calculate(d =>
            {
                var dieFrequency = GroupDieByFrequency(d);
                return dieFrequency.ContainsValue(3) && dieFrequency.ContainsValue(2) ? Scores.HouseScore : 0;
            }, dice, onSuccess, onError);

        public static void LargeStraight(
            int[] dice,
            Action<int> onSuccess,
            Action<string> onError)
            => Calculate(d => d
                .OrderBy(x => x)
                .Zip(
                    d.OrderBy(x => x).Skip(1),
                    (a, b) => b - a
                ).All(diff => diff == 1)
                ? Scores.LargeStraightScore
                : 0, dice, onSuccess, onError);

        public static void SmallStraight(
            int[] dice,
            Action<int> onSuccess,
            Action<string> onError)
            => Calculate(d =>
            {
                var sortedDice = string.Concat(d.OrderBy(x => x).Distinct());
                return IsSmallStraight(sortedDice) ? 30 : 0;
            }, dice, onSuccess, onError);

        private static bool IsSmallStraight(string diceString)
            => diceString.Contains("1234") || diceString.Contains("2345") || diceString.Contains("3456");

        private static bool HasNOfAKind(IEnumerable<int> dice, int n)
            => GroupDieByFrequency(dice).Values.Any(count => count >= n);

        public static void Chance(
            int[] dice,
            Action<int> onSuccess,
            Action<string> onError) => Calculate(d => d.Sum(), dice, onSuccess, onError);

        private static Dictionary<int, int> GroupDieByFrequency(IEnumerable<int> dice)
            => dice.GroupBy(x => x).ToDictionary(g => g.Key, g => g.Count());

        private static void Calculate(
            Func<List<int>, int> compute,
            int[] dice,
            Action<int> onSuccess,
            Action<string> onError)
        {
            if (ValidateRoll(dice, onError))
                onSuccess(compute(dice.ToList()));
        }

        #region Validation

        private static bool ValidateRoll(int[] dice, Action<string> onError)
        {
            if (HasInvalidLength(dice))
            {
                onError("Invalid dice... A roll should contain 5 dice.");
                return false;
            }

            if (ContainsInvalidDie(dice))
            {
                onError("Invalid die value. Each die must be between 1 and 6.");
                return false;
            }

            return true;
        }

        private static bool HasInvalidLength(int[] dice) => dice is not {Length: RollLength};

        private static bool ContainsInvalidDie(IEnumerable<int> dice) => !dice.All(IsValidDie);

        private static bool IsValidDie(int die) => die is >= MinimumDie and <= MaximumDie;

        #endregion

        private static class Scores
        {
            public const int YahtzeeScore = 50;
            public const int HouseScore = 25;
            public const int LargeStraightScore = 40;
        }
    }
}